### Day Objective
- Data Cleaning and Manipulation using Pandas

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/nagamounika5/Datasets/master/Global%20Dataset/Market_Fact.csv')
df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

### Handling Missing or Null values
- Mainly 4 methods to identify and manipulate missing values
    1. isnull(): Presence of missing values, returns boolean data
    2. notnull(): Opposite to isnull()
    3. dropna(): To delete null values
    4. fillna(): Replace null values with any other value

**To identify null values**
- isnull()
- notnull()

In [3]:
df.shape

(8399, 10)

In [4]:
df.isnull()

Ord_id  Prod_id  Ship_id  Cust_id  Sales  Discount  Order_Quantity  \
0      False    False    False    False  False     False           False   
1      False    False    False    False  False     False           False   
2      False    False    False    False  False     False           False   
3      False    False    False    False  False     False           False   
4      False    False    False    False  False     False           False   
5      False    False    False    False  False     False           False   
6      False    False    False    False  False     False           False   
7      False    False    False    False  False     False           False   
8      False    False    False    False  False     False           False   
9      False    False    False    False  False     False           False   
10     False    False    False    False  False     False           False   
11     False    False    False    False  False     False           False   
12     False    False    False    False  False     False           False   
13     False    False    False    False  False     False           False   
14     False    False    False    False  False     False           False   
15     False    False    False    False  False     False           False   
16     False    False    False    False  False     False           False   
17     False    False    False    False  False     False           False   
18     False    False    False    False  False     False           False   
19     False    False    False    False  False     False           False   
20     False    False    False    False  False     False           False   
21     False    False    False    False  False     False           False   
22     False    False    False    False  False     False           False   
23     False    False    False    False  False     False           False   
24     False    False    False    False  False     False           False   
25     False    False    False    False  False     False           False   
26     False    False    False    False  False     False           False   
27     False    False    False    False  False     False           False   
28     False    False    False    False  False     False           False   
29     False    False    False    False  False     False           False   
...      ...      ...      ...      ...    ...       ...             ...   
8369   False    False    False    False  False     False           False   
8370   False    False    False    False  False     False           False   
8371   False    False    False    False  False     False           False   
8372   False    False    False    False  False     False           False   
8373   False    False    False    False  False     False           False   
8374   False    False    False    False  False     False           False   
8375   False    False    False    False  False     False           False   
8376   False    False    False    False  False     False           False   
8377   False    False    False    False  False     False           False   
8378   False    False    False    False  False     False           False   
8379   False    False    False    False  False     False           False   
8380   False    False    False    False  False     False           False   
8381   False    False    False    False  False     False           False   
8382   False    False    False    False  False     False           False   
8383   False    False    False    False  False     False           False   
8384   False    False    False    False  False     False           False   
8385   False    False    False    False  False     False           False   
8386   False    False    False    False  False     False           False   
8387   False    False    False    False  False     False           False   
8388   False    False    False    False  False     False           False   
8389   False    False    False    False  False     Fal

In [5]:
# Columns wise null values
df.isnull().sum()

Ord_id                  0
Prod_id                 0
Ship_id                 0
Cust_id                 0
Sales                   0
Discount                0
Order_Quantity          0
Profit                  0
Shipping_Cost           0
Product_Base_Margin    63
dtype: int64

In [6]:
# total count of null values
df.isnull().sum().sum()

63

*Out of 8399 rows only 63 null values are there. If we drop null values, we can loss very small amount of data*

**Delete null values**
- dropna()

In [7]:
df1 = df.copy()

In [8]:
df1.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [11]:
df1.shape

(8399, 10)

In [12]:
df1.isnull().sum()

Ord_id                  0
Prod_id                 0
Ship_id                 0
Cust_id                 0
Sales                   0
Discount                0
Order_Quantity          0
Profit                  0
Shipping_Cost           0
Product_Base_Margin    63
dtype: int64

- axis = 0 --> Rows
- axis = 1 --> Columns
- inplace = False(no modification will done)
- inplace = True(modification will done in dataset)

In [13]:
df1.dropna(axis = 0, inplace = True)

In [14]:
df1.shape

(8336, 10)

In [15]:
df1.isnull().sum()

Ord_id                 0
Prod_id                0
Ship_id                0
Cust_id                0
Sales                  0
Discount               0
Order_Quantity         0
Profit                 0
Shipping_Cost          0
Product_Base_Margin    0
dtype: int64

**Replace Null values**
- If column is numerical, Replace null values with mean or 0.
- If column is categorical, Replace null values with most occured values or any name.
    - fillna()

In [16]:
df2 = df.copy()

In [17]:
df2.head(3)

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59

In [18]:
df2.shape

(8399, 10)

In [19]:
df2.isnull().sum()

Ord_id                  0
Prod_id                 0
Ship_id                 0
Cust_id                 0
Sales                   0
Discount                0
Order_Quantity          0
Profit                  0
Shipping_Cost           0
Product_Base_Margin    63
dtype: int64

In [20]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
Ord_id                 8399 non-null object
Prod_id                8399 non-null object
Ship_id                8399 non-null object
Cust_id                8399 non-null object
Sales                  8399 non-null float64
Discount               8399 non-null float64
Order_Quantity         8399 non-null int64
Profit                 8399 non-null float64
Shipping_Cost          8399 non-null float64
Product_Base_Margin    8336 non-null float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.2+ KB


In [23]:
m = df2['Product_Base_Margin'].mean()
m

0.5125131957773527

In [24]:
df2['Product_Base_Margin'] = df2['Product_Base_Margin'].fillna(m)

In [25]:
df2.shape

(8399, 10)

In [27]:
df2.isnull().sum().sum()

0

**Delete unwanted columns**

In [28]:
df3 = df.copy()

In [29]:
df3.shape

(8399, 10)

In [30]:
df3.head(3)

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59

In [32]:
df3.drop(['Ord_id'], axis = 1, inplace = True)

In [33]:
df3.head(3)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity   Profit  \
0  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   -30.51   
1  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13     4.56   
2   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26  1148.90   

   Shipping_Cost  Product_Base_Margin  
0           3.60                 0.56  
1           0.93                 0.54  
2           2.50                 0.59

In [34]:
dict1 = {'Students':['Dhakshitha','Lakshmi','Vivek','Siva','Supraja'],
         'Branch': ['CSE','CSE', np.NaN, 'Civil','Mech'],
         'Year': [1, 2, 2, 3, np.NaN,], 
         'Score': [np.NaN, 10, 10, 11, 10]}
data = pd.DataFrame(dict1)

In [35]:
data

Students Branch  Year  Score
0  Dhakshitha    CSE   1.0    NaN
1     Lakshmi    CSE   2.0   10.0
2       Vivek    NaN   2.0   10.0
3        Siva  Civil   3.0   11.0
4     Supraja   Mech   NaN   10.0

In [36]:
data.isnull().sum()

Students    0
Branch      1
Year        1
Score       1
dtype: int64

In [37]:
data.isnull()

Students  Branch   Year  Score
0     False   False  False   True
1     False   False  False  False
2     False    True  False  False
3     False   False  False  False
4     False   False   True  False

In [38]:
data.isnull() == True

Students  Branch   Year  Score
0     False   False  False   True
1     False   False  False  False
2     False    True  False  False
3     False   False  False  False
4     False   False   True  False

In [39]:
data[data.isnull() == True]

Students Branch  Year  Score
0      NaN    NaN   NaN    NaN
1      NaN    NaN   NaN    NaN
2      NaN    NaN   NaN    NaN
3      NaN    NaN   NaN    NaN
4      NaN    NaN   NaN    NaN

In [40]:
data[data['Branch'].isnull() == True]

Students Branch  Year  Score
2    Vivek    NaN   2.0   10.0

In [41]:
help(df.fillna)

Help on method fillna in module pandas.core.frame:

fillna(value=None, method=None, axis=None, inplace=False, limit=None, downcast=None, **kwargs) method of pandas.core.frame.DataFrame instance
    Fill NA/NaN values using the specified method.
    
    Parameters
    ----------
    value : scalar, dict, Series, or DataFrame
        Value to use to fill holes (e.g. 0), alternately a
        dict/Series/DataFrame of values specifying which value to use for
        each index (for a Series) or column (for a DataFrame). (values not
        in the dict/Series/DataFrame will not be filled). This value cannot
        be a list.
    method : {'backfill', 'bfill', 'pad', 'ffill', None}, default None
        Method to use for filling holes in reindexed Series
        pad / ffill: propagate last valid observation forward to next valid
        backfill / bfill: use NEXT valid observation to fill gap
    axis : {0 or 'index', 1 or 'columns'}
    inplace : boolean, default False
        If True, fi

In [42]:
data

Students Branch  Year  Score
0  Dhakshitha    CSE   1.0    NaN
1     Lakshmi    CSE   2.0   10.0
2       Vivek    NaN   2.0   10.0
3        Siva  Civil   3.0   11.0
4     Supraja   Mech   NaN   10.0

In [43]:
data['Branch'].value_counts()

CSE      2
Mech     1
Civil    1
Name: Branch, dtype: int64

In [51]:
data['Branch'] = data['Branch'].fillna('CSE')

In [52]:
data

Students Branch  Year  Score
0  Dhakshitha    CSE   1.0   10.0
1     Lakshmi    CSE   2.0   10.0
2       Vivek    CSE   2.0   10.0
3        Siva  Civil   3.0   11.0
4     Supraja   Mech   3.0   10.0

In [46]:
data['Year'] = data['Year'].fillna(method = 'ffill')

In [47]:
data

Students Branch  Year  Score
0  Dhakshitha    CSE   1.0    NaN
1     Lakshmi    CSE   2.0   10.0
2       Vivek    CSE   2.0   10.0
3        Siva  Civil   3.0   11.0
4     Supraja   Mech   3.0   10.0

In [48]:
data['Score'].fillna(method = 'bfill', inplace = True)

In [49]:
data

Students Branch  Year  Score
0  Dhakshitha    CSE   1.0   10.0
1     Lakshmi    CSE   2.0   10.0
2       Vivek    CSE   2.0   10.0
3        Siva  Civil   3.0   11.0
4     Supraja   Mech   3.0   10.0

In [50]:
data.isnull().sum()

Students    0
Branch      0
Year        0
Score       0
dtype: int64

### Encoding Technique
- Label Encoder
    - To convert categorical data to numerical data

In [53]:
t = pd.read_csv('https://raw.githubusercontent.com/nagamounika5/Datasets/master/titanic.csv')
t.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [54]:
# Check total count of null values

In [55]:
t.shape

(891, 12)

In [56]:
t.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [57]:
t.isnull().sum().sum()

866

In [60]:
t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


In [58]:
from sklearn.preprocessing import LabelEncoder

In [59]:
le = LabelEncoder()

In [61]:
t['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [62]:
t['Sex'] = le.fit_transform(t['Sex'])

In [63]:
t.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S

In [64]:
t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null int32
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int32(1), int64(5), object(4)
memory usage: 80.1+ KB


In [65]:
t['Sex'].value_counts()

1    577
0    314
Name: Sex, dtype: int64

In [66]:
t.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**Task**
- Replace all null values in "Age" column with mean()
- Replace all null values in "Embarked" column with most occured value
- After replacing "Embarked" column, convert object to numerical using Label Encoder